athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsSchemaUtils.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.arrow.dataset.scanner.ScanOptions;
+import org.apache.arrow.dataset.scanner.Scanner;
+import org.apache.arrow.dataset.source.Dataset;
+import org.apache.arrow.dataset.source.DatasetFactory;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.ipc.ArrowReader;
+import org.apache.arrow.vector.types.TimeUnit;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.amazonaws.athena.connectors.gcs.GcsUtil.isFieldTypeNull;
+import static com.amazonaws.athena.connectors.gcs.storage.StorageConstants.BLOCK_PARTITION_COLUMN_NAME;
+import static com.amazonaws.athena.connectors.gcs.storage.StorageUtil.createUri;
+
+public class GcsSchemaUtils
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsSchemaUtils.class);
+
+    private static final ArrowType.ArrowTypeID TIMESTAMP_NANO_TYPE = new ArrowType.Timestamp(TimeUnit.NANOSECOND, null).getTypeID();
+
+    private GcsSchemaUtils()
+    {
+    }
+
+    /**
+     * Builds the table schema based on the provided field by the retrieved instance of {@link StorageTable}
+     *
+     * @param datasource   An instance of {@link StorageMetadata}
+     * @param databaseName Name of the bucket in GCS
+     * @param tableName    Name of the storage object (file) from GCS
+     * @return An instance of {@link Schema}
+     */
+    public static Schema buildTableSchema(StorageMetadata datasource, String databaseName, String tableName) throws Exception
+    {
+        SchemaBuilder schemaBuilder = SchemaBuilder.newBuilder();
+        Optional<StorageTable> optionalStorageTable = datasource.getStorageTable(databaseName, tableName);
+        if (optionalStorageTable.isPresent()) {
+            StorageTable table = optionalStorageTable.get();
+            LOGGER.info("Schema Fields\n{}", table.getFields());
+            for (Field field : table.getFields()) {
+                if (isFieldTypeNull(field)) {
+                    field = Field.nullable(field.getName(), Types.MinorType.VARCHAR.getType());
+                }
+                schemaBuilder.addField(getCompatibleField(field));
+            }
+            schemaBuilder.addStringField(BLOCK_PARTITION_COLUMN_NAME);
+            Schema schema = schemaBuilder.build();
+            return schema;
+        }
+        else {
+            LOGGER.error("Table '{}' was not found under schema '{}'", tableName, databaseName);
+            throw new GcsConnectorException("Table '" + tableName + "' was not found under schema '" + databaseName + "'");
+        }
+    }
+
+    public static Field getCompatibleField(Field field)
```

Comment:
Same here, why are we doing anything with types here? Shouldn't the Dataset API just gives us the correct Arrow Schema Fields?

Reply:
Sometimes it returns an incompatible type per Athena specification. For example, Time (NONOSECONDS) isn't compatible with Athena. We make such a field compatible with Athena and return in such a case.